fix: migrate wiki pages, fix broken links, and clean up Apport docs (#705)
ubuntu/ubuntu-project-docs: docs/contributors/debugging/error-tracker.md

(error-tracker)=
# Error tracker

```{admonition} Migrated content
This page has been migrated from the Ubuntu Wiki "as is".
It may contain inaccuracies, may not be up-to-date, and may require
formatting, language, or other corrections.
```

Ubuntu’s error tracker explains crashes, freezes, and other severe errors to end users; lets them report an error with the click of a button; and collects these reports from millions of users to show Ubuntu developers [which errors are most common](https://errors.ubuntu.com/). It’s all open source, and **you can help**. You can find the errors your computer has reported to the error tracker via System Settings, Security & Privacy, Diagnostics, Show Previous Reports.

> (rationale)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

Statistics collected by Microsoft show that a bug reported by their Windows Error Reporting system “is 4.5 to 5.1 times more likely to be fixed than a bug reported directly by a human”, that fixing the right 1 percent of bugs addresses 50 percent of customer issues, and that fixing 20 percent of bugs addresses 80 percent of customer issues.

The client interface for the error tracker also serves a purpose which is less important for developers, but *more* important for end users: **explaining why something weird just happened**. In previous Ubuntu release versions, when most programs crashed there was no explanation of why the window had disappeared.

(client-design)=
## Client design

(settings)=

> (privacy-settings)=

Reviewer: Is the double anchor okay here?

### Privacy settings

> _Compare [the equivalent screen in the installer](https://docs.google.com/document/d/1bZ4yQIVgGaUGSYu3qiUHnQt3ieBZoqunP_DcleHCr3I/edit#heading=h.3qe17agyh876).

Reviewer: Conversion typo:

```suggestion
Compare [the equivalent screen in the installer](https://docs.google.com/document/d/1bZ4yQIVgGaUGSYu3qiUHnQt3ieBZoqunP_DcleHCr3I/edit#heading=h.3qe17agyh876).
```

(Error reports are also [accessible to trusted Ubuntu developers](https://lists.ubuntu.com/archives/ubuntu-devel-announce/2013-May/001039.html) who are not employed by Canonical. This is covered in the privacy policy.)

The “When on Wi-Fi” and “On any data connection” radio items and caption (bug Bug:1320973) should have exactly the same wording and order as [in the “Updates” settings](https://wiki.ubuntu.com/SoftwareUpdates#settings-phone), but be an independent setting.

![phone-settings-diagnostics-previous-errors.png](error-tracker-images/phone-settings-diagnostics-previous-errors.png)

The “Previous error reports”, “Previous info reports”, and “Previous Wi-Fi reports” screens should all consist of a chronological list of reports, latest first, each labelled by its timestamp. The screen for each report should have that timestamp as its header, and consist entirely of a frame of wrapped selectable and copyable text.

> (error)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

(when-there-is-an-error)=
### When there is an error

**Silent error types** are types of error for which users should not be notified. They consist of:

* Errors that occur during logout/shutdown.

* Any error that occurs over a minute before the error alert could appear, where you have interacted with the session during that interval (bug Bug:1018117). For example, if a CPU-intensive application task crashes while you aren’t using the computer, when you return to the computer you should still see an explanation of why the program has disappeared, regardless of how long that explanation took to appear. But if you were using the computer at the time, you are likely to have noticed the error long before the alert would appear, so the alert would seem like a non-sequitur.

When there is an error that prevents login, and “Send a report automatically if a problem prevents login” is checked, the error should be sent automatically.

Otherwise, if the error is not of a silent type, an alert should appear with controls depending on the situation.

> (multiple)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

![error-alert-layout.png](error-tracker-images/error-alert-layout.png)

![error-alert-expanded-layout.png](error-tracker-images/error-alert-expanded-layout.png)

The “Report previous internal errors too” checkbox should be present if any errors of silent types have occurred since the last time a non-silent error was presented. Whenever it is checked, the details field should show details of all the errors, with separators between them.

If your admin has blocked error reporting, neither the “Send an error report to help fix this problem” nor “Report previous internal errors too” checkboxes should be present. If they are present, they should be checked by default in a new Ubuntu installation, and their state should persist across errors and across Ubuntu sessions.

> (comments)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

![bug-report.png](error-tracker-images/bug-report.png)

**Future work:** Ensure that if there is a delay in displaying a crash, we adjust the text of the dialog to reflect this. As an example, if X crashes and the user has to log in again or reboot the computer.

**Future work:** If a software update is known to fix the problem, replace the primary alert with [the software update alert](https://wiki.ubuntu.com/SoftwareUpdates#alert) (or progress window, depending on the update policy), with customized primary text.  Or point them at a web page (not a wiki page!) with details if a workaround exists, but no fix is available yet.

**Future work:**  Automate the communication with the user to facilitate things like leak detection in subsequent runs, without requiring additional interaction with the user.  Our current process requires us to ask people who are subscribed to the bug to try a specially-instrumented build, with a traditionally very long feedback loop between the developer and the bug subscribers.  We should make it entirely automatic.  Just wait for the next user who sees the bug to click one "yes, I'd like to help make this product better" button.

> (memory)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

(when-there-is-not-enough-memory-for-a-core-dump)=
### When there is not enough memory for a core dump

When the kernel does not have enough memory, an error report should still be sent (or queued for sending) as normal for accounting purposes, just without the core dump.

> (updates)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

If it does not do this, *or* within five seconds it does not receive a response that the problem is fixed by a software update *and* it did not know from a previous submission that it is fixed by a software update, then the error alert should appear as normal.

Otherwise, the usual error alert should change:

* the secondary text should be “The good news is, a software update is available to fix this problem.”

* the checkbox label should be “Send an error report anyway”.

* An extra “Install Updates…” button should be present on the leading side of the trailing group (even if you are not an admin).

![app-crash-reportable-updateable.png](error-tracker-images/app-crash-reportable-updateable.png)

Choosing “Install Updates…” should [launch Software Updater](https://wiki.ubuntu.com/SoftwareUpdates#launch-manual) (leaving the application closed, if it was an application crash).

_We also considered changing the Software Updater UI to appear unprompted sooner, or to have custom text, when updates are known to fix problems users on your system have submitted. We decided against it because it would be less obvious.

> (debconf)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

(when-there-is-a-debconf-prompt)=
### When there is a debconf prompt

debconf prompts for user-installed software in Ubuntu are, overwhelmingly, programming mistakes. Therefore, they should be presented as error alerts.

> **Implementation:** See [/Contributing/Debconf](https://help.ubuntu.com/community//Contributing/Debconf) for instructions on how to work on these alerts.

Reviewer: Replace the broken link with the GitHub archive:

```suggestion
**Implementation:** See [/Contributing/Debconf](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-Contributing-Debconf.wiki) for instructions on how to work on these alerts.
```

(debconf-note)=
**Type “note”**

![debconf-note.png](error-tracker-images/debconf-note.png)

(debconf-text)=
**Type “text” or “error”**

![debconf-text.png](error-tracker-images/debconf-text.png)

(debconf-password)=
**Type “password”**

![debconf-password.png](error-tracker-images/debconf-password.png)

(debconf-progress)=
(presenting-debconf-progress)=
#### Presenting debconf progress

_This has nothing to do with error tracking, but is included here because the error tracker provides all the rest of debconf’s graphical UI.

When a maintainer script requests progress presentation (`db_progress`), the progress text and proportion should be shown in a progress window. Ideally, this progress window should morph to and/or from any consecutive debconf prompts.

![debconf-progress.png](error-tracker-images/debconf-progress.png)

The window title should be of the form ‘Installing “package-name”’, ‘Reinstalling “package-name”’, ‘Updating “package-name”’, or “Removing ‘package-name’” as appropriate, or “Ubuntu” if the type of operation is unknown. The “Skip” button should be present if the operation is skippable.

> (metrics)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

(invitation-for-metrics-collection)=
### Invitation for metrics collection

For any administrator, after the *first* time only that they respond to an error alert, a second alert should appear to invite them to opt in to metrics collection. (The “Esc” key should activate “Don’t Send” in this alert, but the “Enter” key should not do anything.)

![privacy-settings-alert.png](error-tracker-images/privacy-settings-alert.png)

The “Privacy…” button should open System Settings to the Privacy panel. Choosing “Send” should be equivalent to checking “Send occasional system information to Canonical” in the Privacy settings.

> (previous)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

This finally gives us the **average daily error rate** — the number of error reports received that day, divided by the count of weighted active machines for that day.

One problem that isn’t solved by these adjustments: Counting calendar days is misleading. For example, the current error rate for Ubuntu 12.04 is higher on weekdays than in weekends, almost certainly because it’s used on more machines for longer periods during weekdays (bug Bug:1046269).

(comparisons)=
(cross-release-comparisons)=
### Cross-release comparisons

So that we can make fair comparisons of reliability between Ubuntu releases, both the client and the server should track whether each reported error is of a kind that *would* have been reported if it had occurred in each past Ubuntu release.

For example:

* **When there is not enough memory for a coredump**, the error is not recorded in 12.04 or 12.10, but is recorded in 13.04 and later. So errors of that type should be excluded from by-12.04-standards measurements.

* Errors in binaries mentioned in `/etc/apport/blacklist.d` are not counted in 12.04 or 12.10 (bug Bug:1064395).

> (server)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

(errors-ubuntu-com)=

> ## errors.ubuntu.com

Reviewer: If we don't want the heading to render as a link:

```suggestion
## errors\.ubuntu\.com
```

→

![extra-info-approved.png](error-tracker-images/extra-info-approved.png)

Once anyone has entered an apparently valid hook and chosen “Propose”, the hook field should become read-only and the parameter fields should become static text, with “Proposed by:” and “Signed off by:” data appearing below.

The “Reject” and “Needs Work” buttons should be sensitive for anyone signed in with permission to review hooks, and the “Sign Off” button sensitive for any of those people except the initial submitter and someone who has already signed off on this hook.

Once a hook is approved, the “Collecting up to:” line should include a link of the form “3 so far” to the data collected from this hook so far, and “over the next: {duration}” should change to “finishing in: {duration remaining}”.

![extra-info-change.png](error-tracker-images/extra-info-change.png)

If you choose “Propose a Change”, or anyone has submitted a change proposal that has not yet been rejected or signed off, the collection form should begin with radio buttons “Current collection” and “Proposed change”, letting you flip between the hook code *and* other parameters for the current hook vs. the proposed one.

> (server-architecture)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

## Server architecture

> [/ServerArchitecture](https://help.ubuntu.com/community//ServerArchitecture) has additional details.

Reviewer: Replace the broken link with the GitHub archive:

```suggestion
[/ServerArchitecture](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-ServerArchitecture.wiki) has additional details.
```

> (contributing)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

> (how-you-can-help)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

 6. If requested, `whoopsie` sends up the core dump and Daisy writes this to disk (NFS) then puts the path to this file on a Rabbit queue for retracing.

 7. One of the retracers will then pick this core dump off the queue and process it into a [stack trace](http://en.wikipedia.org/wiki/Stack_trace). The stack trace is then computed into a [crash signature](http://bazaar.launchpad.net/~apport-hackers/apport/trunk/view/head:/apport/report.py#L1151) which the database uses as a unique identifier for this problem.

 8. Daisy then increments a counter for the bucket this crash signature belongs to, indicating the number of users who have experienced instances of this problem.

 9. The bucket counts are read and displayed on the Django [http://errors.ubuntu.com](http://errors.ubuntu.com) website.

(how-can-i-test-this)=
### How can I test this?

To see this in action for yourself, simply send any process the SEGV signal:

```bash
eog & sleep 5 && pkill -SEGV eog
```

To see your own reports, load the error tracker page for your uuid:

```bash
xdg-open https://errors.ubuntu.com/user/`sudo cat /var/lib/whoopsie/whoopsie-id`
```

> (get-the-code)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

### Get the code

```bash
mkdir -p ~/bzr
cd ~/bzr
bzr branch lp:daisy
bzr branch lp:errors
bzr branch lp:whoopsie
bzr branch lp:apport
bzr branch lp:~daisy-pluckers/oops-repository/trunk oops-repository.daisy-pluckers
```

(find-bugs-to-fix)=
### Find bugs to fix

* We track all bugs against the [Ubuntu Error Tracker project](https://bugs.launchpad.net/ubuntu-error-tracker).

* Errors that occur within the server-side infrastructure are reported [here](https://errors.ubuntu.com/oops-local/).

> (further-reading)=

Reviewer: An extremely generic ID that probably works as a wiki page anchor but will clash with IDs on other pages in the docs.

### Further reading

* [UDS Raring talk](https://www.youtube.com/watch?v=PPQ7k0jRUE4#t=30m10s)

> * [/Contributing/Errors](https://help.ubuntu.com/community//Contributing/Errors)

Reviewer: ```suggestion
* [/Contributing/Errors](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-Contributing-Errors.wiki)
```

> * [/MapReduce](https://help.ubuntu.com/community//MapReduce)

Reviewer: ```suggestion
* [/MapReduce](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-MapReduce.wiki)

> * [/ServerSideHooks](https://help.ubuntu.com/community//ServerSideHooks)

Reviewer: ```suggestion
* [/ServerSideHooks](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-ServerSideHooks.wiki)
```

> * [/PhasedUpdates](https://help.ubuntu.com/community//PhasedUpdates)

Reviewer: This page has different versions on the two old wikis:

```suggestion
* [/PhasedUpdates](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuCommunityHelpWiki/P/PhasedUpdates.wiki) and [/ErrorTracker/PhasedUpdates](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-PhasedUpdates.wiki)
```

> * [/Deployment](https://help.ubuntu.com/community//Deployment) - How to set up a private error tracker in Juju

Reviewer: ```suggestion
* [/Deployment](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-Deployment.wiki) - How to set up a private error tracker in Juju
```

> * [/Monitoring](https://help.ubuntu.com/community//Monitoring) - Monitoring the production Error Tracker

Reviewer: ```suggestion
* [/Monitoring](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-Monitoring.wiki) - Monitoring the production Error Tracker
```

* [Outstanding code reviews](https://code.launchpad.net/~daisy-pluckers/+activereviews)

> * [/UbuntuReleasePreparation](https://help.ubuntu.com/community//UbuntuReleasePreparation)

Reviewer: I can't find this page in the GitHub archive.

> * [/BreakpadApplicationSupport](https://help.ubuntu.com/community//BreakpadApplicationSupport)

Reviewer: ```suggestion
* [/BreakpadApplicationSupport](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-BreakpadApplicationSupport.wiki)
```

> * [/AutomatedTesting](https://help.ubuntu.com/community//AutomatedTesting)

Reviewer: ```suggestion
* [/AutomatedTesting](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/A/AutomatedTesting.wiki)
```

> * [/CassandraQueries](https://help.ubuntu.com/community//CassandraQueries) - Writing fast queries against Cassandra

Reviewer: ```suggestion
* [/CassandraQueries](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-CassandraQueries.wiki) - Writing fast queries against Cassandra
```

> * [/DailyTasks](https://help.ubuntu.com/community//DailyTasks) - Things to do every day

Reviewer: ```suggestion
* [/DailyTasks](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-DailyTasks.wiki) - Things to do every day
```

> * [/PackageInstallationFailures](https://help.ubuntu.com/community//PackageInstallationFailures)

Reviewer: ```suggestion
* [/PackageInstallationFailures](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-PackageInstallationFailures.wiki)
```

> * [/Statistics](https://help.ubuntu.com/community//Statistics) - Fun with txstatsd and Graphite.

Reviewer: ```suggestion
* [/Statistics](https://github.com/ubuntu/wiki-archives/blob/main/UbuntuWiki/E/ErrorTracker-Statistics.wiki) - Fun with txstatsd and Graphite.
```
